### PR TITLE
Feature: programmatic track selection

### PR DESCRIPTION
**Pull request by joornby-angel**
Our company is migrating from one video player to bitmovin, and we have stored the user's preferred audio and subtitle track selection in our application.  We need these api's in order to keep backwards compatibility with our user's currently selected tracks.

### CODE REVIEW

**Comment by joornby-angel** [comment]
I'm working with my company on the license agreement.  I'll be back in touch shortly.

**Comment by joornby-angel** [comment]
> I'm working with my company on the license agreement. I'll be back in touch shortly.

I signed the CLA, but the status isn't updating.  @pogist @matamegger @rolandkakonyi Is there anything y'all can do to check on it?

**Comment by joornby-angel** [comment]
> > I'm working with my company on the license agreement. I'll be back in touch shortly.
> 
> I signed the CLA, but the status isn't updating. @pogist @matamegger @rolandkakonyi Is there anything y'all can do to check on it?

Just kidding, I realized my error.  Over to y'all for review.  Let me know what I should change or fix.

File: ios/Event+JSON.swift [hunk 1]
```diff
@@ -131,6 +131,38 @@ extension PlayEvent: TimedEventType {}
 extension PausedEvent: TimedEventType {}
 extension PlayingEvent: TimedEventType {}
 
+extension AudioAddedEvent {
+    func toJSON() -> [AnyHashable: Any] {
+        [
+            "name": name,
+            "timestamp": timestamp,
+            "audioTrack": RCTConvert.audioTrackJson(audioTrack)
```
Comment (rolandkakonyi) [inline ios/Event+JSON.swift:139]: Property `time` is missing here

```suggestion
            "audioTrack": RCTConvert.audioTrackJson(audioTrack),
            "time": time
```

File: ios/Event+JSON.swift [hunk 2]
```diff
@@ -131,6 +131,38 @@ extension PlayEvent: TimedEventType {}
...
+        ]
+    }
+}
+
+extension AudioRemovedEvent {
+    func toJSON() -> [AnyHashable: Any] {
+        [
+            "name": name,
+            "timestamp": timestamp,
+            "audioTrack": RCTConvert.audioTrackJson(audioTrack)
```
Comment (rolandkakonyi) [inline ios/Event+JSON.swift:149]: Property `time` is missing here

```suggestion
            "audioTrack": RCTConvert.audioTrackJson(audioTrack),
            "time": time
```

File: ios/PlayerModule.swift [hunk 3]
```diff
@@ -346,4 +386,29 @@ class PlayerModule: NSObject, RCTBridgeModule {
...
+
+    /**
+     Set `nativeId`'s player subtitle track.
+     - Parameter nativeId: Target player Id.
+     - Parameter trackIdentifier: The subtitle track identifier.
+     - Parameter resolver: JS promise resolver.
+     - Parameter rejecter: JS promise rejecter.
+     */
+    @objc(setSubtitleTrack:trackIdentifier:resolver:rejecter:)
+    func setSubtitleTrack(
```
Comment (rolandkakonyi) [inline ios/PlayerModule.swift:398]: Thanks for adding this one retroactively!

**Review by rolandkakonyi — COMMENTED** [review]
Thanks for the contribution! Manual testing looks good!

Please address my comments on the code!

**Comment by rolandkakonyi** [comment]
I found that when using the new sample for track selection on tvOS there are no playback controls are appearing.
@pogist do you maybe know a trick to make them show up? I would assume it is about the focus again.

**Review by matamegger — APPROVED** [review]
Looks good once @rolandkakonyi comments are resolved 🙂

**Comment by rolandkakonyi** [comment]
@joornby-angel please re-request review once my comments get addressed!

**Comment by rolandkakonyi** [comment]
Still open to get feedback from @pogist about the tvOS issue I mentioned in https://github.com/bitmovin/bitmovin-player-react-native/pull/41#issuecomment-1305707115

**Comment by pogist** [comment]
> Still open to get feedback from @pogist about the tvOS issue I mentioned in [#41 (comment)](https://github.com/bitmovin/bitmovin-player-react-native/pull/41#issuecomment-1305707115)

Apparently, this is an issue with the `AVPlayerViewController` itself, not the focus engine (or this PR's code).

Basically, if the `AVPlayerViewController` view is not the same size as the screen, then the system UI controls won’t appear. And it seems there’s no way to opt-out from this behavior when using the system UI.

See more here:

https://developer.apple.com/forums/thread/30458?answerId=101786022#101786022
https://developer.apple.com/forums/thread/19526?answerId=61950022#61950022

**Review by pogist — APPROVED** [review]

File: ios/PlayerModule.swift [hunk 4]
```diff
@@ -347,6 +387,31 @@ class PlayerModule: NSObject, RCTBridgeModule {
...
+        _ nativeId: NativeId,
+        trackIdentifier: String,
```
Comment (rolandkakonyi) [inline ios/PlayerModule.swift:400]: Shouldn't `trackIdentifier` be optional? It is optional for for Android in `PlayerModule.kt`.
In that case the `if` statement can be replaced with a simple `self?.players[nativeId]?.setSubtitle(trackIdentifier: trackIdentifier)` call.

File: src/player.ts [hunk 5]
```diff
@@ -320,13 +321,34 @@ export class Player extends NativeInstance<PlayerConfig> {
...
    * @returns An array containing SubtitleTrack objects for all available subtitle tracks.
    */
   getAvailableSubtitles = async (): Promise<SubtitleTrack[]> => {
     return PlayerModule.getAvailableSubtitles(this.nativeId);
   };
 
+  /**
+   * Sets the subtitle track to the ID specified by trackIdentifier. A list can be retrieved by calling getAvailableSubtitles.
+   */
+  setSubtitleTrack = async (trackIdentifier: string): Promise<void> => {
```
Comment (rolandkakonyi) [inline src/player.ts:348]: I think `trackIdentifier` should be optional. Our public API defines `setSubtitleTrack` with optional identifier parameter.

**Comment by rolandkakonyi** [comment]
@joornby-angel friendly reminder about my last comments, as only those block this feature to be merged.

**Review by rolandkakonyi — COMMENTED** [review]
@joornby-angel Only a minor comment. Other than that it looks great!

**Comment by joornby-angel** [comment]
@rolandkakonyi I accepted your indentation fix, back over to you for more feedback or merging.

**Review by rolandkakonyi — APPROVED** [review]

**Pull request merged**